fix: merge working hours when adjacent

emrysal opened the pull request:
# Fix missing time slots in overlapping ranges + O(n) performance optimization

## Summary

This PR fixes a bug where overlapping availability ranges with the same end time were losing earlier time slots, and optimizes the performance from O(n²) to O(n) to handle large datasets (2000+ ranges) efficiently.

**Original Issue**: When processing availability ranges like `6:00-10:00` and `8:00-10:00`, the system was only showing 2 slots instead of the expected 4, losing the 6:00 and 7:00 slots.

**Changes Made**:
1. **Added failing test case** to demonstrate the bug with overlapping ranges that have the same end time
2. **Fixed merging logic** to properly handle overlapping ranges by taking the earliest start time
3. **Performance optimization** replaced O(n²) `Object.keys().find()` approach with O(n) Map-based lookup for ranges grouped by end time

**Performance Impact**: For event types with 2000 date ranges, this reduces operations from 4M to linear time while maintaining identical merging behavior.

## Review & Testing Checklist for Human

- [ ] **Map state consistency**: Verify `endTimeToKeyMap` stays synchronized with `results` object when ranges are added, merged, or deleted
- [ ] **Algorithmic correctness**: Test with complex overlapping scenarios to ensure O(n) optimization produces identical results to original O(n²) approach  
- [ ] **Performance validation**: Test with large datasets (1000+ availability ranges) to confirm performance improvement and memory usage
- [ ] **Edge case handling**: Test with timezone changes, DST transitions, and complex multi-day availability scenarios
- [ ] **End-to-end testing**: Create event types with many overlapping availability ranges and verify slot generation works correctly in the UI

**Recommended Test Plan**: Create an event type with 20+ overlapping availability ranges including the specific case (6:00-10:00, 8:00-10:00) and verify all expected time slots appear in the booking interface.

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    Input["Availability Input<br/>(WorkingHours[])"]
    BuildDateRanges["buildDateRanges()<br/>packages/lib/date-ranges.ts"]:::context
    ProcessWorkingHours["processWorkingHours()<br/>packages/lib/date-ranges.ts"]:::major-edit
    Results["Record&lt;number, DateRange&gt;<br/>(results object)"]:::context
    EndTimeMap["Map&lt;number, number[]&gt;<br/>(endTimeToKeyMap)"]:::major-edit
    TestFile["date-ranges.test.ts<br/>Added failing test case"]:::minor-edit
    
    Input --> BuildDateRanges
    BuildDateRanges --> ProcessWorkingHours
    ProcessWorkingHours --> Results
    ProcessWorkingHours --> EndTimeMap
    TestFile -.-> ProcessWorkingHours
    
    EndTimeMap --> |"O(1) lookup by end time"| ProcessWorkingHours
    Results --> |"Range merging logic"| ProcessWorkingHours
    
    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Notes

- **Session**: Requested by @emrysal in Devin session: https://app.devin.ai/sessions/755b592417d846cf87c86195d5763197
- **CI Status**: All 37 checks passed including type checking, linting, unit tests, integration tests, and E2E tests
- **Backward Compatibility**: Zero breaking changes - optimization maintains identical output behavior
- **Memory Trade-off**: Uses additional Map storage for O(1) lookups, but provides significant performance gains for high-volume event types

**Critical Note**: While CI passes, this performance optimization introduces new algorithmic complexity that requires careful human validation with realistic datasets to ensure correctness and performance benefits.

emrysal committed:
fix: Adjacency issue when working hours connect over multiple days

emrysal committed:
Merge branch 'main' into bugfix/merge-working-hours-when-adjacent

emrysal committed:
Add tests to validate the new merging of day end logic

emrysal committed:
Merge branch 'main' into bugfix/merge-working-hours-when-adjacent

emrysal committed:
Update to correct annotation.

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

emrysal committed:
Implement subsequent date ranges for date overrides also

emrysal committed:
Merge branch 'bugfix/merge-working-hours-when-adjacent' of https://github.com/calcom/cal.com into bugfix/merge-working-hours-when-adjacent

emrysal committed:
The map needs to be updated on successful resolve.

emrysal committed:
Merge branch 'main' into bugfix/merge-working-hours-when-adjacent

Udit-takkar reviewed (commented):
> Also build this for date overrides (OOO are only full days, 00:00 - 23:59, so already working)

@emrysal  are you working on the above task or raise a follow up PR later?

emrysal committed:
Merge branch 'main' into bugfix/merge-working-hours-when-adjacent

CarinaWolli reviewed (changes requested):
I found one issue that I think was introduced with this change:

If we have two overlapping date ranges like this with the same end time:  
<img width="274" height="109" alt="Screenshot 2025-07-29 at 11 06 37 AM" src="https://github.com/user-attachments/assets/55f51b12-7a80-4e30-b7ed-354205cb4e2a" />

Then we are losing the first date range (6:00-8:00 is missing)
<img width="286" height="162" alt="Screenshot 2025-07-29 at 11 07 01 AM" src="https://github.com/user-attachments/assets/5e6e540c-d51d-4978-962b-50b07823ccd8" />

devin-ai-integration[bot] committed:
test: add failing test for overlapping ranges with same end time

Demonstrates bug where overlapping working hour ranges (6:00-10:00 and 8:00-10:00)
lose the earlier portion (6:00-8:00), showing only 8:00 and 9:00 slots
instead of all 4 slots (6:00, 7:00, 8:00, 9:00).

Related to Carina's comment on PR #21912.

Co-Authored-By: alex@cal.com <me@alexvanandel.com>

devin-ai-integration[bot] committed:
fix: properly merge overlapping ranges with same end time

Fixes bug where overlapping working hour ranges with the same end time
(e.g., 6:00-10:00 and 8:00-10:00) would lose the earlier portion of the
first range. The merging logic now correctly preserves the earliest
start time when ranges overlap and share the same end time.

This ensures all expected time slots are available (6:00, 7:00, 8:00, 9:00)
instead of losing the earlier slots (6:00, 7:00).

Resolves the issue identified in Carina's comment on PR #21912.

Co-Authored-By: alex@cal.com <me@alexvanandel.com>

devin-ai-integration[bot] committed:
perf: optimize overlapping range detection from O(n²) to O(n)

Replaces Object.keys().find() with Map-based lookup for ranges with same end time.
This optimization handles 2000+ date ranges efficiently, reducing complexity from
4M operations to linear time while maintaining the same merging behavior.

Performance improvement for high-volume event types with many availability ranges.

Co-Authored-By: alex@cal.com <me@alexvanandel.com>

CarinaWolli reviewed (changes requested):
Found a bug, that's how you can reproduce it:

Availability:
<img width="582" height="114" alt="Screenshot 2025-07-29 at 4 32 13 PM" src="https://github.com/user-attachments/assets/f65bb2f1-f0c0-448a-bcf7-93aa5d9820e3" />

Monday 3PM slot is available, as it should be: 
<img width="725" height="167" alt="Screenshot 2025-07-29 at 4 32 49 PM" src="https://github.com/user-attachments/assets/4d105d10-1369-4890-a081-fd37147697e6" />


When booking it throws and error:
<img width="817" height="492" alt="Screenshot 2025-07-29 at 4 32 24 PM" src="https://github.com/user-attachments/assets/ddbc475d-0a88-4ef0-bf98-f8260d029f48" />

emrysal commented:
Evaluated the issue reported by @CarinaWolli and it's already an issue; tracked here: https://github.com/calcom/cal.com/issues/22808

anikdhabal committed:
Merge branch 'main' into bugfix/merge-working-hours-when-adjacent
